rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -2366,6 +2366,49 @@ rmw_destroy_topic_names_and_types(
 }
 
 rmw_ret_t
+rmw_get_node_names(
+  const rmw_node_t * node,
+  rmw_node_names_t * node_names)
+{
+  if (!node) {
+    RMW_SET_ERROR_MSG("null node handle");
+    return RMW_RET_ERROR;
+  }
+  if (rmw_check_zero_rmw_node_names(node_names) != RMW_RET_OK) {
+    return RMW_RET_ERROR;
+  }
+
+  // Get participant pointer from node
+  if (node->implementation_identifier != eprosima_fastrtps_identifier) {
+    RMW_SET_ERROR_MSG("node handle not from this implementation");
+    return RMW_RET_ERROR;
+  }
+
+  CustomParticipantInfo * impl = static_cast<CustomParticipantInfo *>(node->data);
+  Participant * participant = impl->participant;
+
+  auto participant_names = participant->getParticipantNames();
+  node_names->node_count = participant_names.size();
+  node_names->names = static_cast<char **>(rmw_allocate(node_names->node_count * sizeof(char *)));
+  for (size_t i = 0; i < participant_names.size(); ++i) {
+    node_names->names[i] = __local_strdup(participant_names[i].c_str());
```

Comment:
The assumption is that `rmw_allocate` should be able to provide a custom strategy (e.g. use memory from a preallocated pool to avoid any allocation at runtime). If any code directly allocates memory (like `strdup` does) it undermines that approach. E.g. this would immediately break any attempt to achieve realtime.

Therefore, yes, I think this should use `rmw_allocate` and `memcpy`.